Text Generation WebUI/src/order-helpers.ts

```diff
@@ -0,0 +1,197 @@
+import {
+  Encryption,
+  EncryptionKey,
+  ResourceType,
+  StorageProviderResource,
+} from '@super-protocol/dto-js';
+import {
+  BlockchainConnector,
+  Crypto,
+  Order,
+  OrderStatus,
+  RIGenerator,
+  getStorageProvider,
+  helpers,
+} from '@super-protocol/sdk-js';
+import fs from 'fs';
+import { Logger } from 'pino';
+import { PassThrough, Readable } from 'stream';
+import tar from 'tar-stream';
+import zlib from 'zlib';
+import { config } from './config';
+import { TunnelProvisionerOrderResult } from './types';
+
+export type DownloadOrderResultParams = {
+  orderId: string;
+  orderKey: EncryptionKey;
+  logger: Logger;
+};
+
+type EncryptedOrderResult = { resource: Encryption; encryption: Encryption };
+
+const isOrderFinished = async (orderId: string): Promise<boolean> => {
+  const order = new Order(orderId);
+
+  if (!(await order.isExist())) {
+    throw new Error(`Order does not exist (orderId=${orderId})`);
+  }
+
+  const orderInfo = await order.getOrderInfo();
+
+  const finishStatuses = [OrderStatus.Done, OrderStatus.Error, OrderStatus.Canceled];
+
+  return finishStatuses.includes(orderInfo.status);
+};
+
+const getOrderEncryptedResult = async (
+  orderId: string,
+  logger: Logger,
+): Promise<EncryptedOrderResult> => {
+  logger.debug('Getting order encrypted result');
+  const connector = BlockchainConnector.getInstance();
+
+  try {
+    await connector.initialize({
+      blockchainUrl: config.blockchainUrl,
+      contractAddress: config.blockchainContractAddress,
+    });
+
+    const retryIntervalInSeconds = 10;
+    await helpers.tryWithInterval({
+      handler() {
+        logger.debug('Checking if order is finished');
+        return isOrderFinished(orderId);
+      },
+      checkResult: (isResultOk) => ({ isResultOk }),
+      retryInterval: retryIntervalInSeconds * 1000,
+      retryMax: (60 * 60 * 3) / retryIntervalInSeconds,
+    });
+
+    const order = new Order(orderId);
+    const orderInfo = await order.getOrderInfo();
+
+    if (orderInfo.status === OrderStatus.Canceled) {
+      throw new Error(`Tunnel Provisioner Order is canceled (orderId=${orderId})`);
+    }
+    if (orderInfo.status === OrderStatus.Error) {
+      throw new Error(`Tunnel Provisioner Order failed (orderId=${orderId})`);
+    }
+
+    const { encryptedResult } = await order.getOrderResult();
+
+    logger.debug('Order encrypted result received');
+
+    return JSON.parse(encryptedResult) as EncryptedOrderResult;
+  } finally {
+    connector.shutdown();
+  }
+};
+
+const decryptOrderResult = async (params: {
```

Comment:
move to sdk. this logic is already used here, in spctl & marketplace

Reply:
My answer - https://github.com/Super-Protocol/solutions/pull/58/files#r1814385047